src/mixin/pending-container.js

```diff
@@ -0,0 +1,57 @@
+
+export const PendingContainer = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			hasPendingChildren: {
+				type: Boolean
+			},
+			pendingCount: {
+				type: Number
+			},
+		};
+	}
+
+	constructor() {
+		super();
+		this.pendingCount = 0;
+		this.hasPendingChildren = false;
+	}
+
+	connectedCallback() {
+		this.addEventListener('pending-state', this.__onPendingState);
+		if (super.connectedCallback) {
+			super.connectedCallback();
+		}
+	}
+
+	disconnectedCallback() {
+		this.addEventListener('pending-state', this.__onPendingState);
+		if (super.disconnectedCallback) {
+			super.disconnectedCallback();
+		}
+	}
+
+	async __onPendingState(e) {
+		const promise = e.detail.promise;
+		if (promise) {
+			this.hasPendingChildren = true;
+			this.pendingCount++;
+			try {
+				await promise;
+			} catch (e) {
+				// to supress uncaught exception logs
+			} finally {
+				this.pendingCount--;
+				if (this.pendingCount === 0) {
+					this.hasPendingChildren = false;
+					const pendingEvent = new CustomEvent('pending-resolved', {
```

Comment:
Just thinking about how this component will be used...

Is the `pending-resolved` event just for testing purposes?  It appears (in the sample below) that re-render is triggered by a change to `hasPendingChildren`.  It looks like `pendingCount` _could_ be private, and doesn't really need to be a property.  Thoughts?

Reply:
Yeh the `pending-resolved` event was my cheap shot at making this more testable....

And yeh, had not given too much thought to the usage of the pending fields.

I think I misread Justin's initial Typescript and thought he had made them both properties, but looking again it looks like only the `hasPendingChildren` is a property and he has annotated it as protected, so I guess it should become:

```
static get properties() {
		return {
			_hasPendingChildren: {
				type: Boolean
			}
		};
	}
```

and as you say pendingCount can just be a private field.